src/types/felt252_dict.rs

```diff
@@ -28,22 +34,79 @@ pub fn build<'ctx>(
     context: &'ctx Context,
     _module: &Module<'ctx>,
     _registry: &ProgramRegistry<CoreType, CoreLibfunc>,
-    _metadata: &mut MetadataStorage,
-    _info: WithSelf<InfoAndTypeConcreteType>,
+    metadata: &mut MetadataStorage,
+    info: WithSelf<InfoAndTypeConcreteType>,
 ) -> Result<Type<'ctx>> {
+    metadata
+        .get_or_insert_with::<SnapshotClonesMeta>(SnapshotClonesMeta::default)
+        .register(
+            info.self_ty().clone(),
+            snapshot_take,
+            InfoAndTypeConcreteType {
+                info: info.info.clone(),
+                ty: info.ty.clone(),
+            },
+        );
+
     Ok(llvm::r#type::pointer(context, 0))
 }
 
+#[allow(clippy::too_many_arguments)]
+fn snapshot_take<'ctx, 'this>(
+    context: &'ctx Context,
+    _registry: &ProgramRegistry<CoreType, CoreLibfunc>,
+    entry: &'this Block<'ctx>,
+    location: Location<'ctx>,
+    helper: &LibfuncHelper<'ctx, 'this>,
+    metadata: &mut MetadataStorage,
+    _info: WithSelf<InfoAndTypeConcreteType>,
+    src_value: Value<'ctx, 'this>,
+) -> Result<(&'this Block<'ctx>, Value<'ctx, 'this>)> {
+    // This snapshot take assumes the value types in a felt are always copy and don't need deep cloning.
+    // So far (atleast up to 2.6.4) you cannot make a felt dict of array values, so this holds.
```

Comment:
and what about structs?

Reply:
it can hold arbitrary types via a Nullable<T> value. So if we have a struct `A` (or an array, another dict, etc), `FeltDict252<Nullable<A>>` is valid.

Reply:
This comment contradicts the findings in #723. Indeed, you can create and take snapshots from a `DictFelt<Array<Felt>>`.